sky/provision/aws/config.py

```diff
@@ -222,10 +223,17 @@ def _get_role(role_name: str):
 
 
 @functools.lru_cache(maxsize=128)  # Keep bounded.
-def _get_route_tables(ec2, vpc_id: Optional[str], main: bool) -> List[Any]:
+def _get_route_tables(ec2, vpc_id: Optional[str], region: str,
```

Comment:
Does this mean the `lru_cache` does not compare the first argument `ec2` when deciding whether to rerun the function?

Reply:
Just tested it with a minimal test and it seems indeed different ec2 objects are not considered different arguments:

```
from sky.adaptors import aws
import functools

@functools.lru_cache(maxsize=10)
def diff_ec2(ec2):
    print('called')

ec2 = aws.resource('ec2', region_name='us-east-1')
ec2_2 = aws.resource('ec2', region_name='us-west-1')

diff_ec2(ec2)
diff_ec2(ec2_2)
```

```
python ./test-lru.py 
called
```
```

Reply:
It does compare the `ec2` property but that object is the same one passed in every time so essentially does nothing  for the cache. I confirmed this by debugging and looking at `ec2.__hash__()`. This is the same for every call to the function. So before my change there was only `main: bool` that could cause a different value. Adding region will cache based on region and main now which I believe is what we want.

Reply:
> Just tested it with a minimal test and it seems indeed different ec2 objects are not considered different arguments:
> 
> ```python
> from sky.adaptors import aws
> import functools
> 
> @functools.lru_cache(maxsize=10)
> def diff_ec2(ec2):
>     print('called')
> 
> ec2 = aws.resource('ec2', region_name='us-east-1')
> ec2_2 = aws.resource('ec2', region_name='us-west-1')
> 
> diff_ec2(ec2)
> diff_ec2(ec2_2)
> ```
> 
> ```shell
> python ./test-lru.py 
> called
> ```

For this test to actually work you should print out the current time in milliseconds instead of 'called'. If the times are the same you got cache, if they are different you did not get cache.

Reply:
> Just tested it with a minimal test and it seems indeed different ec2 objects are not considered different arguments:
> 
> ```python
> from sky.adaptors import aws
> import functools
> 
> @functools.lru_cache(maxsize=10)
> def diff_ec2(ec2):
>     print('called')
> 
> ec2 = aws.resource('ec2', region_name='us-east-1')
> ec2_2 = aws.resource('ec2', region_name='us-west-1')
> 
> diff_ec2(ec2)
> diff_ec2(ec2_2)
> ```
> 
> ```shell
> python ./test-lru.py 
> called
> ```

Strange, what is the hash of both of these objects? `ec2.__hash--()` and `ec2_2.__hash__()`

Reply:
yeah sorry that is correct. That __hash__ function on the ec2 object must not be very well written then :-)